doc/manual/src/glossary.md

```diff
@@ -29,6 +51,17 @@
     store object by downloading a pre-built version of the store object
     from some server.
 
+  - [substituter]{#gloss-substituter}\
+    A *substituter* is a store other than `/nix/store` from which Nix will
+    copy a store path instead of building it.  Nix will not copy a store
+    path from a remote store unless one of the following is true:
+
+    - the store object is signed by one of the `trusted-public-keys`
+    - the substituter is in the `trusted-substituters` list
+    - the `no-require-sigs` option has been set to disable signature checking
+    - the store object is a derivation
+    - the store object is the realisation of a fixed-output derivation
```

Comment:
This is a laundry list of special cases, we should break down the algorithm so it makes more sense.

- Content-addressed data is just data, and always "safe" to share. Derivations and fixed output derivations' outputs are just examples of content-addressed data.
- input-addressed things (everything is content-addressed or input-addressed) need at least one signature that corresponds to a trusted public key, unless we disabled this with `no-require-sigs`.

Reply:
Incidentally, this is a a great example of why @fricklerhandwerk's architecture docs are so important.

If we go straight into how user-facing knobs work, we have to either explain the data model on the fly (awkward diversion) or do a laundry list of familiar special cases (hard to remember, ad-hoc). But if we have a solid reference explaining the data model, we can hyperlink it instead, and not have to re-explain things or do laundry list. Yay!

Reply:
It is possible to sign content-addressed things, but this is an historical accident and it would be better if it didn't exist. In an ideal all-CA world, one only signs the "trust maps" which *relate* derviations and output.

Both the derivations and outputs themselves are plain old data and thus inherently trustworthy. Only the relationship between them is an unproven claim that needs to be authenticated with signatures.

Reply:
> This is a laundry list of special cases, we should break down the algorithm so it makes more sense.

What about `trusted-substituters`?

> * Derivations and fixed output derivations' outputs are just examples of content-addressed data.

Thanks, updated in f20a421c3 and then un-inlined in cb5ffeada.

I left the definition of "content-addressed derivation" as a FIXME because I'm not sure exactly how it should read.  Do you have a suggestion?  The obvious "has a `__contentAddressed=true` attr" conflicts with "all FODs are CADs", so I guess I'm not clear on the exact definition.

> a laundry list of familiar special cases (hard to remember, ad-hoc).

I don't think that's what this is.  It is meant to be an exhaustive list of non-overlapping conditions.  It is a security-critical detail; it needs to be specified somewhere, and at the moment it isn't.

The very simple question "when does Nix require a signature on something being added to the store" isn't answered anywhere.

> But if we have a solid reference explaining the data model, we can hyperlink it instead, and not have to re-explain things or do laundry list.

That's great; I look forward to the day when we have that and it provides a simple and direct answer to this important question.